src/client/event_handler.rs

```diff
@@ -10,38 +10,91 @@ use crate::model::application::interaction::Interaction;
 use crate::model::guild::automod::{ActionExecution, Rule};
 use crate::model::prelude::*;
 
-/// The core trait for handling events by serenity.
-#[async_trait]
-pub trait EventHandler: Send + Sync {
+macro_rules! event_handler {
+    ( $(
+        $( #[doc = $doc:literal] )*
+        $( #[cfg(feature = $feature:literal)] )?
+        async fn $method_name:ident(&self, $variant_name:ident { $(
+            $arg_name:ident: $arg_type:ty
+        ),* $(,)? } );
+    )* ) => {
+        /// The core trait for handling events by serenity.
+        #[async_trait]
+        pub trait EventHandler: Send + Sync { $(
+            $( #[doc = $doc] )* $( #[cfg(feature = $feature)] )?
+            async fn $method_name(&self, $( $arg_name: $arg_type ),* ) {
+                // Suppress unused argument warnings
+                drop(( $( $arg_name ),* ));
+            }
+        )* }
+
+        /// This enum stores every possible event that an [`EventHandler`] can receive.
+        #[non_exhaustive]
+        #[allow(clippy::large_enum_variant)] // TODO: do some boxing to fix this
+        pub enum FullEvent { $(
+            $( #[doc = $doc] )* $( #[cfg(feature = $feature)] )?
+            $variant_name {
+                $( $arg_name: $arg_type ),*
+            },
+        )* }
+
+        impl FullEvent {
+            /// Returns the name of this event as a snake case string
+            ///
+            /// ```rust,no_run
+            /// # use serenity::client::{Context, FullEvent};
+            /// # fn _foo(ctx: Context, event: FullEvent) {
+            /// if let FullEvent::Message { .. } = &event {
+            ///     assert_eq!(event.snake_case_name(), "cache_ready");
```

Comment:
Shouldn't this be `message` if the event is `FullEvent::Message`?

```suggestion
            ///     assert_eq!(event.snake_case_name(), "message");
```